13893699 Add FactoryGirl for testing

lucyconklin opened the pull request:
- Add to Gemfile
- Apply to me_spec

[#13893699]

lucyconklin committed:
Add FactoryGirl for testing

- Add to Gemfile
- Apply to me_spec

[#13893699]

jasonnoble commented on Gemfile:
```diff
@@ -14,4 +14,5 @@ group :test do
   gem 'vcr'
   gem 'rake'
   gem 'pry'
+  gem "factory_girl", "~> 4.0"
```
Is the ~> required?  Why not just `gem 'factory_girl'`?

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+require './lib/pivotal_tracker'
```
best practice is to have multiple .rb files in spec/factories directory.  So maybe `spec/factories/me.rb`?

lucyconklin replied:
Agreed. Kyle and I did that in the updated PR

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+  factory :me, class: PivotalTracker::Me do
+    api_token '0'
+    created_at DateTime.now
+    email 'example@example.com'
```
`email { Faker::Internet.email(name) }`

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+    created_at DateTime.now
+    email 'example@example.com'
+    has_google_identity true
+    id 1
```
Probably want a auto-incrementing value here?

lucyconklin replied:
@jasonnoble any tips on how to implement? My google search is suggesting a random number from a range. Should this be a sequence?

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+    id 1
+    initals "FL"
+    kind "me"
+    name "username"
```
`name { Faker::Name.name }`

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+    email 'example@example.com'
+    has_google_identity true
+    id 1
+    initals "FL"
```
`initials { name.split.map{|name| name[0] }`  will work, I think.

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+    personal_settings ({})
+    project_ids ["#{ENV['PROJECT_ID']}"]
+    receives_in_app_notifications true
+    time_zone Time.now.zone
```
`time_zone { Time.zone.now }`

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+    receives_in_app_notifications true
+    time_zone Time.now.zone
+    updated_at DateTime.now
+    username "username"
```
`username { Faker::Internet.user_name(name) }`

jasonnoble commented on spec/factories/factories.rb:
```diff
@@ -0,0 +1,25 @@
+    username "username"
+    workspace_ids [1,2]
+
+    initialize_with { new(attributes) }
```
Is this not the default?

lucyconklin replied:
~~No, this was to make sure the attributes are sent through as a hash~~

Just kidding, it works when that line is commented out

jasonnoble commented on spec/lib/pivotal-tracker/me_spec.rb:
```diff
@@ -1,30 +1,15 @@
       expect(me).to be_a(PivotalTracker::Me)
 
-      attributes.each do |attribute, value|
+      # binding.pry
```
Remove me...

lucyconklin committed:
Add factory for membership_summary

- changed the name of factories.rb to reflect object

[#138939699]

lucyconklin committed:
Add following factory

[#138939699]

lucyconklin committed:
Add factory for google_attachment

[#138939699]

lucyconklin committed:
Add iteration factory

[#138939699]

lucyconklin committed:
Add factory for jira_integration

[#138939699]

lucyconklin committed:
Add project snapshot factory

[#138939699]

lucyconklin committed:
Add factory for project_workspace

[#138939699]

lucyconklin committed:
Update Gemfile.lock for Faker gem

[#138939699]

jasonnoble commented on Gemfile:
```diff
@@ -15,4 +15,5 @@ group :test do
   gem 'rake'
   gem 'pry'
   gem 'faker'
+  gem "factory_girl", "~> 4.0"
```
Would `gem 'factory_girl'` work here?

lucyconklin committed:
Add factory for project_summary

- Also implement super condensed tests

[#138939699]

lucyconklin committed:
Refactor tests and require for google_attachemnts

[#138939699]

lucyconklin committed:
Failing tests and half started files

[#138939699]

jasonnoble commented on spec/factories/following.rb:
```diff
@@ -0,0 +1,11 @@
+
+FactoryGirl.define do
+  factory :following, class: PivotalTracker::Following do
+    id        1
```
Should be a https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#sequences

lucyconklin replied:
@jasonnoble I used sequence like this:
`sequence(:id)  { |n| n}`

jasonnoble commented on spec/factories/following.rb:
```diff
@@ -0,0 +1,11 @@
+    person_id 1
+    story_id  1
+    epic_id   1
+    kind      'kind'
```
`'following'` according to the docs

jasonnoble commented on spec/lib/pivotal-tracker/following_spec.rb:
```diff
@@ -1,16 +1,9 @@
-    end
+  it_behaves_like "an API backed model" do
+    let(:attributes) { FactoryGirl.attributes_for(:following) }
+    let(:subject) { PivotalTracker::Following.new(attributes) }
```
`model.new(attributes)`  #DRY

lucyconklin replied:
@jasonnoble I am going to do this now in all the specs and wrap it into one commit

kheppenstall committed:
Change require statement to just file in following factory

[#138939699]

kheppenstall committed:
Create iteration override factory

[#138939699]

kheppenstall committed:
File attachment factory

-Change file attachment spec to general api backed model form
-Create file attachment factory

[#138939699]

kheppenstall committed:
Update iteration factory

[#138939699]

kheppenstall committed:
Update jira integration factory

[#138939699]

kheppenstall committed:
Change require statement to just needed file in jira integration factory

[#138939699]

kheppenstall committed:
Update membership summary factory

[#138939699]

kheppenstall committed:
Update project snapshot factory

[#138939699]

kheppenstall committed:
Update project summary factory

[#138939699]

kheppenstall committed:
Update project workspace factory

[#138939699]

lucyconklin committed:
Add factory for follower

[#138939699]

lucyconklin committed:
Add factory for notification

[#138939699]

lucyconklin committed:
Add factory for other_integration

[#138939699]

lucyconklin committed:
Use factory_girl gem without version

[#138939699]

lucyconklin committed:
Align model attributes

[#138939699]

lucyconklin committed:
Unfinished factories need to be completed

[#138939699]

lucyconklin committed:
Add factory for external story

[#138939699]

lucyconklin committed:
Add factory for external story

[#138939699]

lucyconklin committed:
Add factory for get_satisfaction_integration

[#138939699]

lucyconklin committed:
Add factory for get_satisfaction_integration

[#138939699]

kheppenstall committed:
Add label factory

-Refactor test to use factory
-Create factory for the label resource

[#138939699]

kheppenstall committed:
Project Id and Version factory

-Create project id and version factory
-Refactor unit test to use factory

[#138939699]

kheppenstall committed:
Merge branch '138939699_add_factory_girl_api_version5' of https://github.com/lucyconklin/pivotal-tracker into 138939699_add_factory_girl_api_version5

kheppenstall committed:
Merge branch '138939699_add_factory_girl_api_version5' of https://github.com/lucyconklin/pivotal-tracker into 138939699_add_factory_girl_api_version5

kheppenstall committed:
Extract subject to spec helper for API backed resource unit test

[#138939699]

lucyconklin committed:
Add id sequence for :id in factories

[#138939699]

Fix spacing in sequenced ids

[#138939699]

jasonnoble commented on spec/factories/follower.rb:
```diff
@@ -0,0 +1,12 @@
+    email         { Faker::Internet.email }
+    initials      { name.split.map{|name| name[0] } }
+    username      { Faker::Internet.user_name }
+    kind          'kind'
```
All of these `kind` values should be what the API returns...  A good practice is to make the factories look like data you would use in the real world.  Developer users of the gem can look to the factories to determine what the data looks like.

lucyconklin replied:
Working on this now.

@jasonnoble let me know if you have suggestions about google_kind in factories/google_attachement.rb researching that now...

tmikeschu replied:
https://www.pivotaltracker.com/help/api/rest/v5#follower_resource

wlffann replied:
Ex. for this one would be `kind: follower`?

![follower](https://cloud.githubusercontent.com/assets/18703294/22600742/19391c3e-e9f9-11e6-9de3-0caf194cf648.jpeg)

lucyconklin replied:
@tmikeschu @wlffann Yes. But for google_attachment I was looking for an acceptable string. 

I ended up using "calendar#calendar".

jasonnoble commented on spec/factories/get_satisfaction_integration.rb:
```diff
@@ -0,0 +1,19 @@
+    updated_at    DateTime.now
+    company       'company'
+    product       'product'
+    kind          'kind'
```
...

lucyconklin committed:
Shorten :id sequence in factories

- take out { |n| n } leaving only sequence(:id)

[#138939699]

lucyconklin committed:
Change factory attributes for :kind to be descriptive

- Assign all kind attributes to conform to Pivotal API documentation

[#138939699]

lucyconklin commented on spec/factories/file_attachment.rb:
```diff
@@ -0,0 +1,20 @@
+    big_url       { Faker::Internet.url }
+    thumbnail_url { Faker::Internet.url }
+    kind          'file_attachment'
+  end
```
@jasonnoble So I changed all of these to be strings, ~~but the docs say that the strings are objects, but it doesn't seem reasonable to store objects in these factories.~~

Upon closer reading I think these strings satisfy the documentation.

From the docs:

string 
 —  The type of this object: file_attachment. This field is read only.

source:
https://www.pivotaltracker.com/help/api/rest/v5#file_attachment_resource

tmikeschu replied:
@lucyconklin I interpret that as a string that declares it's type. AKA I support your move.

<img width="637" alt="screen shot 2017-02-03 at 10 47 35 am" src="https://cloud.githubusercontent.com/assets/19894032/22601879/320fc97e-e9fe-11e6-9119-2252a494880c.png">
